### PR TITLE
Allow Reloading of Custom Permissions

### PR DESCRIPTION
**Pull request by willies952002**
Add the ability to reload the custom permissions file by doing "/reload permissions"
Implements Feature Request: https://github.com/PaperMC/Paper/issues/49

Sorry about reposting/recreating this, but I was unable to re-open the original PR.

### CODE REVIEW

**Comment by willies952002** [comment]
Also... if this doesn't work for you, please let me know what's wrong and I'll try and fix it.

**Comment by willies952002** [comment]
to prevent confusion, that last commit was an amend & force-push :P

File: Spigot-Server-Patches/0085-Allow-Reloading-of-Custom-Permissions.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index cb9906a..c49c022 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -819,7 +819,7 @@ public final class CraftServer implements Server {
+             try {
+                 pluginManager.addPermission(perm);
+             } catch (IllegalArgumentException ex) {
+-                getLogger().log(Level.SEVERE, "Permission in " + file + " was already defined", ex);
++                getLogger().log(Level.WARNING, "Permission in " + file + " was already defined"); // Paper - Just Display as Warning
```
Comment (kashike) [inline Spigot-Server-Patches/0085-Allow-Reloading-of-Custom-Permissions.patch]: I'd say to leave this as `SEVERE`, log messages will change and people might filter by `SEVERE`/`WARNING`/etc.

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+         this.setAliases(Arrays.asList("rl"));
+     }
+ 
+     @Override
+     public boolean execute(CommandSender sender, String currentAlias, String[] args) {
++
+         if (!testPermission(sender)) return true;
+ 
++        // Paper Start - Reload permissions.yml (PAPER-49)
++        if ( args.length == 1 ) {
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: extra spaces

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
++            if ( args[0].equalsIgnoreCase("permissions") || args[0].equalsIgnoreCase("perms") ) {
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: extra spaces

File: Spigot-Server-Patches/0085-Allow-Reloading-of-Custom-Permissions.patch [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+         }
+     }
+@@ -1833,4 +1833,14 @@ public final class CraftServer implements Server {
+     {
+         return spigot;
+     }
++
++    // Paper start
++    @Override
++    public void reloadPermissions() {
```
Comment (kashike) [inline Spigot-Server-Patches/0085-Allow-Reloading-of-Custom-Permissions.patch:19]: This won't work. This needs to be moved to `SimplePluginManager` as:

``` java
   // Paper start
    public void clearPermissions() {
        synchronized (this) {
            this.permissions.clear();
            this.defaultPerms.get(true).clear();
            this.defaultPerms.get(false).clear();
        }
    }
    // Paper end
```

or similar.

and then:

``` java
    @Override
    public void reloadPermissions() {
        ((SimplePluginManager) this.pluginManager).clearPermissions();
        this.loadCustomPermissions();
    }
```

or similar.

**Comment by willies952002** [comment]
@kashike thank you for the input, I will make the appropriate changes now.

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
++
+         Command.broadcastCommandMessage(sender, ChatColor.RED + "Please note that this command is not supported and may cause issues when using some plugins.");
+         Command.broadcastCommandMessage(sender, ChatColor.RED + "If you encounter any issues please use the /stop command to restart your server.");
+         Bukkit.reload();
+@@ -32,7 +44,7 @@ public class ReloadCommand extends BukkitCommand {
+     @Override
+     public java.util.List<String> tabComplete(CommandSender sender, String alias, String[] args) throws IllegalArgumentException
+     {
+-        return java.util.Collections.emptyList();
++        return Arrays.asList("perms"); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: `return java.util.Collections.singletonList("permissions");`

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
++import org.bukkit.Server;
+ import org.bukkit.command.Command;
+ import org.bukkit.command.CommandSender;
+ 
+@@ -11,15 +12,26 @@ public class ReloadCommand extends BukkitCommand {
+     public ReloadCommand(String name) {
+         super(name);
+         this.description = "Reloads the server configuration and plugins";
+-        this.usageMessage = "/reload";
++        this.usageMessage = "/reload [perms|permissions]"; // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: I'd say just choose one, preferably "permissions".

**Comment by willies952002** [comment]
@kashike thank you again for the feedback, will change now.

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+         this.setPermission("bukkit.command.reload");
+         this.setAliases(Arrays.asList("rl"));
+     }
+ 
+     @Override
+     public boolean execute(CommandSender sender, String currentAlias, String[] args) {
++
+         if (!testPermission(sender)) return true;
+ 
++        // Paper Start - Reload permissions.yml (PAPER-49)
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: Start -> start, to be consistent

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
++        if (args.length == 1) {
++            if (args[0].equalsIgnoreCase("permissions") || args[0].equalsIgnoreCase("perms")) {
++                Bukkit.reloadPermissions();
++                Command.broadcastCommandMessage(sender, ChatColor.GREEN + "Permissions successufully reloaded.");
++                return true;
++            }
++        }
++        // Paper End
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: End -> end, to be consistent

**Comment by Ichbinjoe** [comment]
These commits will eventually need rebased to be PR'd

**Comment by willies952002** [comment]
That. Was. A. Nightmare.

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
++
+         Command.broadcastCommandMessage(sender, ChatColor.RED + "Please note that this command is not supported and may cause issues when using some plugins.");
+         Command.broadcastCommandMessage(sender, ChatColor.RED + "If you encounter any issues please use the /stop command to restart your server.");
+         Bukkit.reload();
+@@ -32,7 +47,8 @@ public class ReloadCommand extends BukkitCommand {
+     @Override
+     public java.util.List<String> tabComplete(CommandSender sender, String alias, String[] args) throws IllegalArgumentException
+     {
+-        return java.util.Collections.emptyList();
++        return Collections.singletonList("permissions"); // Paper
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: Don't import it.

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
++    void reloadPermissions(); // Paper
+ }
+diff --git a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+index c70d512..f91063b 100644
+--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
++++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+@@ -1,25 +1,40 @@
+ package org.bukkit.command.defaults;
+ 
++import java.util.ArrayList;
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: ???

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+ import java.util.Arrays;
++import java.util.Collections;
++import java.util.List;
+ 
+ import org.bukkit.Bukkit;
+ import org.bukkit.ChatColor;
+ import org.bukkit.command.Command;
+ import org.bukkit.command.CommandSender;
++import org.bukkit.plugin.Plugin;
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: ???

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+index c70d512..f91063b 100644
+--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
++++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+@@ -1,25 +1,40 @@
+ package org.bukkit.command.defaults;
+ 
++import java.util.ArrayList;
+ import java.util.Arrays;
++import java.util.Collections;
++import java.util.List;
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: ???

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+diff --git a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+index c70d512..f91063b 100644
+--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
++++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+@@ -1,25 +1,40 @@
+ package org.bukkit.command.defaults;
+ 
++import java.util.ArrayList;
+ import java.util.Arrays;
++import java.util.Collections;
```
Comment (kashike) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: ???

Comment (Ichbinjoe) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: Is the synchronized block necessary? Can't we just catch bad calls with a

## sync catcher?

Joseph Hirschfeld
joe@ibj.io
Ichbinjoe
Skype: ichbinjoehirschfeld

File: Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+index f93ca2e..cd73bfa 100644
+--- a/src/main/java/org/bukkit/Bukkit.java
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -1188,4 +1188,11 @@ public final class Bukkit {
+     {
+         return server.spigot();
+     }
++
++    // Paper start
++    public static void reloadPermissions() {
```
Comment (aikar) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: I don't think this one is needed, it's not going to be a common thing, so lets not litter up the Bukkit util class.

Bukkit.getServer().reloadPermissions() is sufficient API.

Comment (willies952002) [inline Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch]: Removed

**Pull request merged**